### PR TITLE
Fix rabbitmq

### PR DESCRIPTION
**Pull request by SteamRiC**

### CODE REVIEW

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 1]
```diff
@@ -12,24 +12,32 @@
...
+        @logger = Logger.new(STDOUT)
         @queue = arguments.fetch(:queue, 'logstash-queue')
         @level = level
-
         @message_builder = MessageBuilder.new(service)
-        @freddy = Freddy.build(logger, config)
+
+        begin
+          @freddy = Freddy.build(@logger, config.merge({recover_from_connection_close: true}))
+        rescue Bunny::Exception
```
Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: what are you catching here?

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 2]
```diff
@@ -12,24 +12,32 @@
...
     class RabbitmqAdapter
-      attr_reader :publisher
+      attr_reader :freddy
 
       CONFIGURATION_KEYS = [:host, :user, :pass, :port]
 
       def initialize(level, service, arguments = {})
         config = arguments.select { |key, value| CONFIGURATION_KEYS.include?(key) }
-        logger = Logger.new(STDOUT)
+        @logger = Logger.new(STDOUT)
```
Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: why instance variable?

File: lib/logasm/adapters/rabbitmq_adapter/publisher.rb [hunk 3]
```diff
@@ -1,20 +0,0 @@
-require 'bunny'
```
Comment (stenver) [inline lib/logasm/adapters/rabbitmq_adapter/publisher.rb]: Can you remove bunny from gemfiles now?

File: lib/logasm/adapters/rabbitmq_adapter.rb (hunk 1 above)
Comment (madis) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: begin - rescue - end is not needed here (just rescue is enough if you're at the top level of a method)

```ruby
def viskaja
  raise StandardError.new
rescue StandardError
  puts "SIIN!"
end
```

Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: this is not proper way to handle errors. if you catch an error here then this class is useless. you can't use anything else

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 4]
```diff
@@ -12,24 +12,32 @@
...
       end
 
       def log(level, metadata = {})
         if meets_threshold?(level)
           message = @message_builder.build_message metadata, level
-          @freddy.deliver @queue, metadata
+
+          begin
+            @freddy.deliver @queue, message
+          rescue Bunny::Exception, NoMethodError => e
```
Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: what are you catching here?

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 5]
```diff
@@ -12,24 +12,32 @@
...
+        rescue Bunny::Exception
+        end
       end
 
       def log(level, metadata = {})
         if meets_threshold?(level)
           message = @message_builder.build_message metadata, level
-          @freddy.deliver @queue, metadata
+
+          begin
```
Comment (madis) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: I would move sending & catching to a separate method that's only responsibility is to call the method that can raise and to rescue.

**Comment by SteamRiC** [comment]
Updated

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 6]
```diff
@@ -21,15 +21,18 @@ def initialize(level, service, arguments = {})
         logger = Logger.new(STDOUT)
         @queue = arguments.fetch(:queue, 'logstash-queue')
         @level = level
-
         @message_builder = MessageBuilder.new(service)
-        @freddy = Freddy.build(logger, config)
+
+        begin
+          @freddy = Freddy.build(logger, config.merge({recover_from_connection_close: true}))
+        rescue Bunny::TCPConnectionFailed => e
```
Comment (urmastalimaa) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: I don't think that this exception should be caught silently as this class is unusable on this path.

Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: +1

File: lib/logasm/adapters/rabbitmq_adapter.rb [hunk 7]
```diff
@@ -38,6 +41,11 @@ def log(level, metadata = {})
       def meets_threshold?(level)
         LOG_LEVELS.index(level.to_s) >= @level
       end
+
+      def deliver_message(message)
+        @freddy.deliver @queue, message
+      rescue Bunny::ConnectionClosedError, NoMethodError => e
```
Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: when is the `NoMethodError` used?

Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: no need for `=> e` if you're not gonna use it

Comment (SteamRiC) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: when freddy.build fails

File: lib/logasm/adapters/rabbitmq_adapter.rb (hunk 6 above)
Comment (SteamRiC) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: freddy has a logger so it will log the error

Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: but this class will be unusable

File: lib/logasm/adapters/rabbitmq_adapter.rb (hunk 7 above)
Comment (indrekj) [inline lib/logasm/adapters/rabbitmq_adapter.rb]: when freddy.build fails then we should not continue using this class. blow up instead

**Comment by indrekj** [comment]
+1

**Pull request merged**